### PR TITLE
20181215 tom

### PR DESCRIPTION
**Pull request by mmpagani**
This PR adds the ability to compute seismic hazard using the classical approach in case of a source cluster, as the one used by the USGS for the calculation of hazard in the New Madrid area. Adding this capability required a major refactoring of the Temporal Occurrence Model (TOM) object which is now defined in a more general way using an abstract class. This transition is not yet complete as some of the TOM methods require an occurrence rate. This is a parameter that should be set at the instantiation level i.e. right after parsing the input model. This is not trivial (nor the main priority at this time). 

The number of occurrences of the cluster now is hardcoded and set to 50. In the future we will consider the possibility of using a check on the accumulator.

This feature must be considered experimental. Additional tests with real cases will be performed once a first implementation will enter into master. 

An example of .xml file describing a cluster source is included as a test.

### CODE REVIEW

File: openquake/hazardlib/tests/source_model/source_group_cluster.xml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+xmlns:gml="http://www.opengis.net/gml"
+>
+    <sourceModel>
+            <sourceGroup
+            name="group 1"
+            tectonicRegion="Active Shallow Crust" 
+            cluster="true">
+            
+                    <poissonTOM occurrence_rate="0.001" />
+
```
[the remark this answers is not in the document]
Comment (mmpagani) [reply openquake/hazardlib/tests/source_model/source_group_cluster.xml]: I also considered the solution you propose. I found it not as flexible as the solution that I finally adopted particularly for cases where more parameters are required. Anyhow I am open to different solutions.

File: openquake/hazardlib/tom.py [hunk 2]
```diff
@@ -29,28 +30,102 @@
...
+        parameter value.
+        """
+
+    @abc.abstractmethod
+    def get_probability_n_occurrences(self):
+        """
+        Calculate the probability of occurrence of a number of events in the
+        constructor's ``time_span``.
+        """
+
```
[the remark this answers is not in the document]
Comment (mmpagani) [reply openquake/hazardlib/tom.py]: Okay agreed, I'll remove the originally defined `get_probability_one_occurrence`

File: openquake/hazardlib/tom.py [hunk 3]
```diff
@@ -61,12 +136,21 @@ def get_probability_one_or_more_occurrences(self, occurrence_rate):
         return 1 - math.exp(- occurrence_rate * self.time_span)
 
     def get_probability_one_occurrence(self, occurrence_rate):
+        return scipy.stats.poisson(occurrence_rate * self.time_span).pmf(1)
+
```
[the remark this answers is not in the document]
Comment (mmpagani) [reply openquake/hazardlib/tom.py]: idem

File: openquake/hazardlib/tests/source_model/source_group_cluster.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (micheles) [reply openquake/hazardlib/tests/source_model/source_group_cluster.xml]: It is true that what I propose is less flexible, but it is the approach we use for the IMTs and works even for multiple parameters, if they are positional and expressible as Python literals (i.e. float, integers, strings and tuple/dict containing literals). It should be enough.

File: openquake/hazardlib/calc/hazard_curve.py [hunk 4]
```diff
@@ -116,8 +134,31 @@ def classical(group, src_filter, gsims, param, monitor=Monitor()):
...
+    # Processing
+    if cluster:
+        first = True
+        pmapclu = AccumDict({grp_id: ProbabilityMap(len(imtls.array), 
+                            len(gsims)) for grp_id in grp_ids})
+        # Get temporal occurrence model
+        tom = param.get('temporal_occurrence_model')
+        # Number of occurrences for the cluster
+        for nocc in range (0, 50):
+            # TODO fix this once the occurrence rate will be used just as
```
Comment (micheles) [inline openquake/hazardlib/calc/hazard_curve.py]: It is not clear to me tha meaning of the range(0, 50): can you comment on that?

File: openquake/hazardlib/sourceconverter.py [hunk 5]
```diff
@@ -801,12 +823,19 @@ def convert_sourceGroup(self, node):
         srcs_weights = node.attrib.get('srcs_weights')
         grp_attrs = {k: v for k, v in node.attrib.items()
                      if k not in ('name', 'src_interdep', 'rup_interdep',
-                                  'srcs_weights')}
+                                  'srcs_weights', 'group_type')}
         sg = SourceGroup(trt, min_mag=self.minimum_magnitude)
```
Comment (micheles) [inline openquake/hazardlib/sourceconverter.py]: What's the purpose of `group_type`? Perhaps a leftover?

File: openquake/hazardlib/sourceconverter.py [hunk 6]
```diff
@@ -801,12 +823,19 @@ def convert_sourceGroup(self, node):
...
         sg.name = node.attrib.get('name')
+        # set attributes related to occurrence
         sg.src_interdep = node.attrib.get('src_interdep', 'indep')
         sg.rup_interdep = node.attrib.get('rup_interdep', 'indep')
         sg.grp_probability = node.attrib.get('grp_probability')
+        # set the cluster attribute
+        sg.cluster = False
+        if node.attrib.get('cluster') == 'true':
+            sg.cluster = True
+        #
```
Comment (micheles) [inline openquake/hazardlib/sourceconverter.py]: You can do this in a single line:
```
sg.cluster = node.attrib.get('cluster') == 'true'
```

**Pull request merged**